### PR TITLE
Build wheels for pdffit2 (non-pure Python) package

### PR DESCRIPTION
**Pull request by bobleesj**
I have tested both full and pre-release of pdffit2 and tested using my Mac Arm64. 

### Test result:

Download from test pypi:

```
 conda activate diffpy.pdffit2_env
pip install -i https://test.pypi.org/simple/ diffpy.pdffit2==1.4.4rc8
Collecting diffpy.pdffit2==1.4.4rc8
```

Pytest result:

```               
tests/test_exceptions.py ........................................................................             [ 63%]
tests/test_pdffit.py ..............................                                                           [ 90%]
tests/test_phase_fractions.py ..                                                                              [ 92%]
tests/test_shape_factors.py ........                                                                          [ 99%]
tests/test_version.py .                                                                                       [100%]
```

### Ref:

- Test PyPI: https://test.pypi.org/project/diffpy.pdffit2/1.4.6/#files
- GitHub Actions: https://github.com/bobleesj/diffpy.pdffit2/actions/runs/11876636736

### CODE REVIEW

File: .github/workflows/_build-pdffit2-package.yml
Comment (bobleesj) [inline .github/workflows/_build-pdffit2-package.yml:1]: Thanks to @Tieqiong.

File: .github/workflows/_build-wheel-release-upload.yml [hunk 1]
```diff
@@ -22,47 +22,35 @@ on:
...
-  tag-check:
-    runs-on: ubuntu-latest
-    outputs:
-      tag_valid: ${{ steps.tag-check.outputs.tag_valid }}
-    steps:
-      - name: Extract and validate tag for (pre)-release
-        id: tag-check
-        run: |
-          TAG_NAME=${GITHUB_REF#refs/tags/}
+  tag-privilege-check:
```
Comment (bobleesj) [inline .github/workflows/_build-wheel-release-upload.yml:25]: Here, I've modularized checking the valid tag and privilege using a separate file `./.github/workflows/_release_tag_privilege_check.yml` that resides in this repository.

File: .github/workflows/_build-wheel-release-upload.yml [hunk 2]
```diff
@@ -22,47 +22,35 @@ on:
...
+    uses: ./.github/workflows/_release_tag_privilege_check.yml
+    
+  build-pure-python-package:
```
Comment (bobleesj) [inline .github/workflows/_build-wheel-release-upload.yml:28]: After tag check, build pure python - again modularize the pure python build

File: .github/workflows/_build-wheel-release-upload.yml [hunk 3]
```diff
@@ -22,47 +22,35 @@ on:
...
+    needs: [tag-privilege-check]
+    if: inputs.c_extension == false
+    uses: ./.github/workflows/_build-pure-python-package.yml
+    
+  build-non-pure-python-package:
+    needs: [tag-privilege-check]
+    if: inputs.project == 'diffpy.pdffit2'
```
Comment (bobleesj) [inline .github/workflows/_build-wheel-release-upload.yml:35]: After tag check and also checking the name `diffpy.pdffit2`, build sdist and wheel files - again modularize the python build . The design here is scalable. We have to provide a build recipe for a package containing C++ code.

**Review by bobleesj — COMMENTED** [review]
Please see in-line comments for your review. Thank you @sbillinge

**Comment by bobleesj** [comment]
(Applied pre-commit)

File: .github/workflows/_build-pdffit2-package.yml
Comment (bobleesj) [inline .github/workflows/_build-pdffit2-package.yml:1]: Thanks to @Tieqiong.

File: .github/workflows/_release_tag_privilege_check.yml
Comment (bobleesj) [inline .github/workflows/_release_tag_privilege_check.yml:1]: Modularized tag and user privilege workflow

File: .github/workflows/_build-wheel-release-upload.yml (hunk 1 above)
Comment (sbillinge) [reply .github/workflows/_build-wheel-release-upload.yml:25]: OK, I like this

**Review by sbillinge — COMMENTED** [review]
wow, this is super work!  Thank you SOOOOOO much @bobleesj and @Tieqiong 

One quick question....for consistency, do we want all the `uses: ..... .yml` to be `uses:... .yml@v0`?

**Comment by bobleesj** [comment]
> wow, this is super work! Thank you SOOOOOO much @bobleesj and @Tieqiong
> 
> One quick question....for consistency, do we want all the `uses: ..... .yml` to be `uses:... .yml@v0`?

@sbillinge Would you be referring to the below `uses` cases below?

```
  tag-privilege-check:
    uses: ./.github/workflows/_release_tag_privilege_check.yml

  build-pure-python-package:
    needs: [tag-privilege-check]
    if: inputs.c_extension == false
    uses: ./.github/workflows/_build-pure-python-package.yml
...
```

For the relative paths above (./.github/...), it appears we aren't allowed to add `@v0` shown below. 

Screenshot 2024-11-17 at 11 26 03 AM

I assume it is because the top very workflow from `pdffit2`  has already checked out `v0`:

```
jobs:
  release:
    permissions:
      contents: write
    uses: Bilingegroup/release-scripts/.github/workflows/_build-wheel-release-upload.yml@v0 # HERE
```

**Comment by sbillinge** [comment]
> > wow, this is super work! Thank you SOOOOOO much @bobleesj and @Tieqiong
> > One quick question....for consistency, do we want all the `uses: ..... .yml` to be `uses:... .yml@v0`?
> 
> @sbillinge Would you be referring to the below `uses` cases below?
> 
> ```
>   tag-privilege-check:
>     uses: ./.github/workflows/_release_tag_privilege_check.yml
> 
>   build-pure-python-package:
>     needs: [tag-privilege-check]
>     if: inputs.c_extension == false
>     uses: ./.github/workflows/_build-pure-python-package.yml
> ...
> ```
> 
> For the relative paths above (./.github/...), it appears we aren't allowed to add `@v0` shown below.
> Screenshot 2024-11-17 at 11 26 03 AM
> 
> I assume it is because the top very workflow from `pdffit2` has already checked out `v0`:
> 
> ```
> jobs:
>   release:
>     permissions:
>       contents: write
>     uses: Bilingegroup/release-scripts/.github/workflows/_build-wheel-release-upload.yml@v0 # HERE
> ```

OK, so I think that means that it will all work as desired.  We will check out `v0` and `v0` branch will contain the right versions of the yml's so we don't need `v0` here, right?  It just don't want any ambiguity that part of the script is using `v0` and another is using `main`, which will sometimes work and sometimes not and therefore cause immense confusion.   Please confirm that my understanding is correct and I can merge this.

**Comment by bobleesj** [comment]
> We will check out v0 and v0 branch will contain the right versions of the yml's so we don't need v0 here

Yes, that's correct. We checkout `v0` and we get even more reusable workflow through the relative paths within `v0`.

**Pull request merged**